### PR TITLE
Implement statistics

### PR DESCRIPTION
**Pull request by MarcosNicolau**
This pr adds vm statistics to be used in the zksync-era integration, that includes:
- Collecting precompiles round counts
- Collect storages reads and writes cycles.
- Collect vm number of instructions run (or, as they call it, monotonic counter)
- Collect decommit cycles

This implementation is supported by the [circuits tracer tests](https://github.com/lambdaclass/zksync-era/pull/243/commits/8abf5c725da6050fdcbfa6eabe0fc7a2856046a5) in the zksync-era integration.

### CODE REVIEW

**Comment by Oppen** [comment]
Great idea! Two comments:
1. I haven't checked, so maybe it's already there but not mentioned in the description, both stats for far and near call depths would be useful.
2. This should probably be gated by a feature flag, since it may add significant overhead.

File: src/precompiles/ecrecover.rs [hunk 1]
```diff
@@ -18,11 +18,13 @@ pub use zkevm_opcode_defs::sha3::Keccak256;
 
 use super::*;
 
+const NUM_ROUNDS: usize = 1;
```
Comment (IAvecilla) [inline src/precompiles/ecrecover.rs:21]: We have this constant defined for two different precompiles, but it has the same value in both cases. We could move it to the `statistics.rs` file and rename it to something like `DEFAULT_PRECOMPILE_NUM_ROUNDS` to have it in one place.

File: src/opcode.rs [hunk 2]
```diff
@@ -1,14 +1,14 @@
+use crate::eravm_error::EraVmError;
+use crate::eravm_error::OpcodeError;
 use lazy_static::lazy_static;
-use zkevm_opcode_defs::Opcode as Variant;
+pub use zkevm_opcode_defs::Opcode as Variant;
```
Comment (IAvecilla) [inline src/opcode.rs:4]: Was this change necessary for the `zksync-era` integration?

File: src/state.rs [hunk 3]
```diff
@@ -268,7 +277,7 @@ impl VMState {
             .map
             .iter()
             .filter_map(|(key, value)| {
-                let initial_value = self.storage_read_inner(key);
+                let initial_value = self.storage.borrow_mut().storage_read(key);
```
Comment (IAvecilla) [inline src/state.rs:280]: This is outside the scope, so it's not necessary to change it here. However, I'll add a comment to link this to a future issue. The `storage_read()` operation should not require a mutable reference.

**Review by IAvecilla — APPROVED** [review]

File: src/opcode.rs (hunk 2 above)
Comment (MarcosNicolau) [reply src/opcode.rs:4]: Yes, as the libraries used diverge, this way we don't have to add a new crate.

File: src/state.rs (hunk 3 above)
Comment (MarcosNicolau) [reply src/state.rs:280]: Noted, I'll make sure to add a comment to ensure we correct this in a future refactor!

**Review by fkrause98 — APPROVED** [review]

**Review by fkrause98 — APPROVED** [review]
